lib/zone.ts

```diff
@@ -1273,613 +1238,51 @@ const Zone: ZoneType = (function(global: any) {
           try {
             task.zone.runTask(task, null, null);
           } catch (error) {
-            consoleError(error);
-          }
-        }
-      }
-      while (_uncaughtPromiseErrors.length) {
-        while (_uncaughtPromiseErrors.length) {
-          const uncaughtPromiseError: UncaughtPromiseError = _uncaughtPromiseErrors.shift();
-          try {
-            uncaughtPromiseError.zone.runGuarded(() => {
-              throw uncaughtPromiseError;
-            });
-          } catch (error) {
-            handleUnhandledRejection(error);
+            if ((Zone as any)[__symbol__('ignoreConsoleErrorUncaughtError')]) {
+              return;
+            }
+            _api.consoleError(showError, error);
           }
         }
       }
+      _api.microtaskDrainDone(showError);
       _isDrainingMicrotaskQueue = false;
     }
   }
 
-  function isThenable(value: any): boolean {
-    return value && value.then;
-  }
+  //////////////////////////////////////////////////////
+  //////////////////////////////////////////////////////
+  ///  BOOTSTRAP
+  //////////////////////////////////////////////////////
+  //////////////////////////////////////////////////////
 
-  function forwardResolution(value: any): any {
-    return value;
-  }
 
-  function forwardRejection(rejection: any): any {
-    return ZoneAwarePromise.reject(rejection);
-  }
-
-  const symbolState: string = __symbol__('state');
-  const symbolValue: string = __symbol__('value');
-  const source: string = 'Promise.then';
-  const UNRESOLVED: null = null;
-  const RESOLVED = true;
-  const REJECTED = false;
-  const REJECTED_NO_CATCH = 0;
-
-  function makeResolver(promise: ZoneAwarePromise<any>, state: boolean): (value: any) => void {
-    return (v) => {
-      try {
-        resolvePromise(promise, state, v);
-      } catch (err) {
-        resolvePromise(promise, false, err);
-      }
-      // Do not return value or you will break the Promise spec.
-    };
-  }
-
-  const once = function() {
-    let wasCalled = false;
+  const NO_ZONE = {name: 'NO ZONE'};
+  const notScheduled: 'notScheduled' = 'notScheduled', scheduling: 'scheduling' = 'scheduling',
+                      scheduled: 'scheduled' = 'scheduled', running: 'running' = 'running',
+                      canceling: 'canceling' = 'canceling', unknown: 'unknown' = 'unknown';
+  const microTask: 'microTask' = 'microTask', macroTask: 'macroTask' = 'macroTask',
+                   eventTask: 'eventTask' = 'eventTask';
 
-    return function wrapper(wrappedFunction: Function) {
-      return function() {
-        if (wasCalled) {
-          return;
-        }
-        wasCalled = true;
-        wrappedFunction.apply(null, arguments);
-      };
-    };
+  const patches: {[key: string]: any} = {};
+  const _api: _ZonePrivate = {
```

Comment:
1) I want to keep in hidden so that you can only get to it from the callback.

2) We could do that, but since we only have one place which needs it so far, I would prefer to have a use case before we complicate things.